src/uct/cuda/base/cuda_iface.h

```diff
@@ -58,20 +58,21 @@
     UCT_CUDADRV_FUNC(_func, UCS_LOG_LEVEL_ERROR)
 
 
-#define UCT_CUDADRV_CTX_ACTIVE(_state)                                        \
-    {                                                                         \
-        CUcontext _cur_ctx;                                                   \
-        CUdevice _dev;                                                        \
-        unsigned _flags;                                                      \
-                                                                              \
-        _state = 0;                                                           \
-        /* avoid active state check if no cuda activity */                    \
-        if ((CUDA_SUCCESS == cuCtxGetCurrent(&_cur_ctx)) &&                   \
-            (NULL != _cur_ctx)) {                                             \
-            UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetDevice(&_dev));                  \
-            UCT_CUDADRV_FUNC_LOG_ERR(cuDevicePrimaryCtxGetState(_dev, &_flags,\
-                                                                &(_state)));  \
-        }                                                                     \
+#define UCT_CUDADRV_CTX_ACTIVE(_state) \
+    { \
+        CUdevice _dev; \
+        CUcontext _ctx; \
+        int _flags; \
+        if (CUDA_SUCCESS == cuCtxGetDevice(&_dev)) { \
+            cuDevicePrimaryCtxGetState(_dev, &_flags, &_state); \
+            if (_state == 0) { \
+                /* need to retain for malloc purposes */ \
+                cuDevicePrimaryCtxRetain(&_ctx, _dev); \
```

Comment:
UCT can't know when the user is done with CUDA runtime use. Also, we can't retain and release here because when refcount on context drops to zero, it is reset and any subsequent call to cuMemAlloc (in cuda_copy UCT) would fail. For now, we are letting cleanup happen as part of the process termination.

Reply:
dont we need to check the return from API if retain is a success?

Reply:
why exactly we are are doing this?  isn't it expected application has a valid context when we allocate the memory?

Reply:
We want to support the case of user calling the following:
```
cudaSetDevice(0)
uct_mem_alloc(cuda_copy_md,...)
```
but the sequence
```
cudaSetDevice(0)
cuMemAlloc(...)
```
fails even though the user has made intention clear about which device to use. For this not to fail, runtime's primary device context must be retained. This is because some runtime API calls like `cudaSetDevice` don't end up setting a context even though device ordinal is populated. [Please refer](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#context)

> dont we need to check the return from API if retain is a success?

Agreed. Will add a check.

Reply:
would it have been automatic if we use cudaMalloc() instead of cuMemAlloc()?